modules/apps/27-interchain-accounts/types/metadata.go

```diff
@@ -27,6 +27,21 @@ func NewMetadata(version, controllerConnectionID, hostConnectionID, accAddress,
 	}
 }
 
+// IsMetadataEqual compares a metadata to a pevious version string set in a channel struct.
+// It ensure all fields are equal except the Address string
+func IsPreviousMetadataEqual(previousVersion string, metadata Metadata) bool {
```

Comment:
Suggestion...

We already have code to unmarhsall the version string [here](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L43-L46) and the same code is found in [`OnChannelOpenTry`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/host/keeper/handshake.go#L41-L44) and [`OnOpenChannelAck`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/controller/keeper/handshake.go#L76-L79). And now we also have it inside this function. So we could make an extra constructor function and remove the duplication of code in all these places. So basically to create a function with a signature like this: `func NewMetadata(version string) (Metadata, error)`.

And if you do that then you could also change this function so that it has the `Metadata` receiver type and change also the name and signature, so something like this: `func (Metadata m) AreEqual(metadata Metadata) bool`. I leave it up to you if you prefer a pointer receiver type.

The you can invoke these functions like:

```
previousMetadata, err := icatypes.NewMetadata(previousVersion)
if !err {
    return "", err
}
if !metadata.AreEqual(previousMetadata) {
    ...   
}

```

Reply:
I like your suggestion. `NewMetadata` would need to be named something else to avoid name collision with the traditional usage of `NewMetadata`. `IsEqual` is also a good idea, but we will remove all this logic with the removal of active channels, so I'm leaning towards leaving as is. It makes the equality check explicitly to be used for active channel security and nothing else (otherwise external users may start depending on that API which should be removed in the future)

Reply:
I prefer @crodriguezvega's suggestion too, but given this will be removed in a couple of months hopefully, then I think we should just leave it as is.

Reply:
agree with @damiannolan